File: bot/services/storage.py
from __future__ import annotations

import os
from datetime import datetime
from pathlib import Path

import aiofiles
from telegram import Bot

STORAGE_ROOT = Path(os.environ.get("STORAGE_PATH", "./uploads")).resolve()


def ensure_storage() -> None:
    STORAGE_ROOT.mkdir(parents=True, exist_ok=True)


def _ext_from_mime(mime: str | None) -> str:
    if not mime:
        return "bin"
    if "jpeg" in mime or "jpg" in mime:
        return "jpg"
    if "png" in mime:
        return "png"
    if "pdf" in mime:
        return "pdf"
    return "bin"


async def download_telegram_file(
    bot: Bot,
    file_id: str,
    telegram_user_id: int,
    doc_type: str,
    mime_type: str | None = None,
) -> str:
    ensure_storage()
    tg_file = await bot.get_file(file_id)
    ext = _ext_from_mime(mime_type)
    ts = datetime.utcnow().strftime("%Y%m%d_%H%M%S")
    user_dir = STORAGE_ROOT / str(telegram_user_id)
    user_dir.mkdir(parents=True, exist_ok=True)
    fname = f"{doc_type}_{ts}.{ext}"
    dest = user_dir / fname
    await tg_file.download_to_drive(custom_path=str(dest))
    return str(dest)

File: src/services/storage.js
import fs from "node:fs/promises";
import path from "node:path";

const STORAGE_ROOT = path.resolve(
  process.env.STORAGE_PATH || path.join(process.cwd(), "uploads")
);

export function ensureStorage() {
  return fs.mkdir(STORAGE_ROOT, { recursive: true });
}

function extFromMime(mime) {
  if (!mime) return "bin";
  if (mime.includes("jpeg") || mime.includes("jpg")) return "jpg";
  if (mime.includes("png")) return "png";
  if (mime.includes("pdf")) return "pdf";
  if (mime.includes("mp4")) return "mp4";
  if (mime.startsWith("video/")) return "mp4";
  return "bin";
}

/** `telegram` — `ctx.telegram` (Telegraf Telegram client), `Telegraf` emas. */
export async function downloadTelegramFile(telegram, fileId, telegramUserId, docType, mimeType) {
  await ensureStorage();
  const link = await telegram.getFileLink(fileId);
  const ext = extFromMime(mimeType);
  const ts = new Date().toISOString().replace(/[-:]/g, "").slice(0, 15).replace("T", "_");
  const userDir = path.join(STORAGE_ROOT, String(telegramUserId));
  await fs.mkdir(userDir, { recursive: true });
  const fname = `${docType}_${ts}.${ext}`;
  const dest = path.join(userDir, fname);
  const res = await fetch(link.href);
  if (!res.ok) {
    throw new Error(`Failed to download file: ${res.status}`);
  }
  const buf = Buffer.from(await res.arrayBuffer());
  await fs.writeFile(dest, buf);
  return dest;
}

export { STORAGE_ROOT };
